src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,401 @@
+            return this;
+        }
+
+        /**
+         * Akin to a {@code Builder.build()} method for an {@code Extract}
+         *
+         * @return an immutable {@code Extract}
+         */
+        public Extract extractOnly() {
+            if (this.ikms.isEmpty() && this.salts.isEmpty()) {
+                throw new IllegalStateException(
+                    "this `Builder` must have either at least one IKM value, at least one salt "
+                    + "value, or values for both before calling `extractOnly`");
+            } else {
+                this.extract = new Extract(List.copyOf(ikms), List.copyOf(salts));
+                return this.extract;
+            }
+        }
+
+        /**
+         * Akin to a {@code Builder.build()} method for an {@code ExtractExpand}
+         *
+         * @param info
+         *     the info
+         * @param length
+         *     the length
+         *
+         * @return an {@code ExtractExpand}
+         */
+        public ExtractExpand andExpand(byte[] info, int length) {
+            if (extract == null) {
+                throw new IllegalStateException(
+                    "`thenExpand` can only be called on a `Builder` when `ofTypeExtract` has "
+                    + "been called");
+            }
+            return extractExpand(new Extract(List.copyOf(ikms), List.copyOf(salts)), info, length);
+        }
+
+        /**
+         * {@code addIKM} may be called when the ikm value is to be assembled piece-meal or if part
+         * of the IKM is to be supplied by a hardware crypto device. This method appends to the
+         * existing list of values or creates a new list if there are none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the IKM but the actual value
+         * of the IKM is not yet available.
+         *
+         * @param ikm
+         *     the ikm value (null values will not be added)
+         *
+         * @return a new {@code Extract} object
+         */
+        public Builder addIKM(SecretKey ikm) {
+            if (ikm != null) {
+                ikms.add(ikm);
+            }
+            return this;
+        }
+
+        /**
+         * {@code addIKM} may be called when the ikm value is to be assembled piece-meal or if part
+         * of the IKM is to be supplied by a hardware crypto device. This method appends to the
+         * existing list of values or creates a new list if there are none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the IKM but the actual value
+         * of the IKM is not yet available.
+         *
+         * @param ikm
+         *     the ikm value (null or empty values will not be added)
+         *
+         * @return a new {@code Extract} object
+         */
+        public Builder addIKM(byte[] ikm) {
+            if (ikm != null && ikm.length != 0) {
+                return addIKM(new SecretKeySpec(ikm, "RAW"));
```

Comment:
Usually "RAW" is for format. I'd rather use "IKM" or "Generic". Same for `addSalt`.